File: B형대비/추가+문제/[실전실습코드] 이승재_성적조회.cpp
#ifndef _CRT_SECURE_NO_WARNINGS
#define _CRT_SECURE_NO_WARNINGS
#endif

#include <stdio.h>

extern void init();
extern int add(int mId, int mGrade, char mGender[7], int mScore);
extern int remove(int mId);
extern int query(int mGradeCnt, int mGrade[], int mGenderCnt, char mGender[][7], int mScore);

/////////////////////////////////////////////////////////////////////////

#define CMD_INIT 100
#define CMD_ADD 200
#define CMD_REMOVE 300
#define CMD_QUERY 400

static bool run() {
	int q;
	scanf("%d", &q);

	int id, grade, score;
	char gender[7];
	int cmd, ans, ret;
	bool okay = false;

	for (int i = 0; i < q; ++i) {
		scanf("%d", &cmd);
		switch (cmd) {
			case CMD_INIT:
				init();
				okay = true;
				break;
			case CMD_ADD:
				scanf("%d %d %s %d %d", &id, &grade, gender, &score, &ans);
				ret = add(id, grade, gender, score);
				if (ans != ret)
					okay = false;
				break;
			case CMD_REMOVE:
				scanf("%d %d", &id, &ans);
				ret = remove(id);
				if (ans != ret)
					okay = false;
				break;
			case CMD_QUERY: {
				int gradeCnt, genderCnt;
				int gradeArr[3];
				char genderArr[2][7];
				scanf("%d", &gradeCnt);
				if (gradeCnt == 1) {
					scanf("%d %d", &gradeArr[0], &genderCnt);
				} else if (gradeCnt == 2) {
					scanf("%d %d %d", &gradeArr[0], &gradeArr[1], &genderCnt);
				} else {
					scanf("%d %d %d %d", &gradeArr[0], &gradeArr[1], &gradeArr[2], &genderCnt);
				}
				if (genderCnt == 1) {
					scanf("%s %d %d", genderArr[0], &score, &ans);
				} else {
					scanf("%s %s %d %d", genderArr[0], genderArr[1], &score, &ans);
				}
				ret = query(gradeCnt, gradeArr, genderCnt, genderArr, score);
				if (ans != ret)
					okay = false;
				break;
			}
			default:
				okay = false;
				break;
		}
	}
	return okay;
}

int main() {
	setbuf(stdout, NULL);
	//freopen("sample_input.txt", "r", stdin);

	int T, MARK;
	scanf("%d %d", &T, &MARK);

	for (int tc = 1; tc <= T; tc++) {
		int score = run() ? MARK : 0;
		printf("#%d %d\n", tc, score);
	}

	return 0;
}
#include <bits/stdc++.h>
using namespace std;
struct student {
	int id, score;
};
bool operator<(const student &a, const student &b) {
	if (a.score != b.score) return a.score < b.score;
	return a.id < b.id;
}
set<student> st[3][2];
unordered_map<int, tuple<int, int, int>> tps;
void init() {
	for (int i = 0; i < 3; ++i) {
		for (int j = 0; j < 2; ++j) {
			st[i][j].clear();
		}
	}
	tps.clear();
	return;
}

int add(int mId, int mGrade, char mGender[7], int mScore) {
	--mGrade;
	auto &au = st[mGrade][mGender[0] == 'f'];
	au.insert({mId, mScore});
	tps[mId] = {mGrade, mGender[0] == 'f', mScore};
	return au.rbegin()->id;
}

int remove(int mId) {
	if (tps.count(mId) == 0) {
		return 0;
	}
	int grade, gender, score;
	tie(grade, gender, score) = tps[mId];
	auto &au = st[grade][gender];
	au.erase({mId, score});
	tps.erase(mId);
	if (au.empty()) return 0;
	return au.begin()->id;
}

int query(int mGradeCnt, int mGrade[], int mGenderCnt, char mGender[][7], int mScore) {
	student ret{0, (int)1e9};
	for (int i = 0; i < mGradeCnt; ++i) {
		--mGrade[i];
		for (int j = 0; j < mGenderCnt; ++j) {
			auto &au = st[mGrade[i]][mGender[j][0] == 'f'];
			auto it = au.lower_bound({0, mScore});
			if (it != au.end()) {
				if (*it < ret) {
					ret = *it;
				}
			}
		}
	}
	return ret.id;
}

File: B형대비/추가+문제/[실전실습코드] 이승재_성적조회.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

import java.util.*;

class Solution {
	private final static int CMD_INIT = 100;
	private final static int CMD_ADD = 200;
	private final static int CMD_REMOVE = 300;
	private final static int CMD_QUERY = 400;

	private final static UserSolution usersolution = new UserSolution();

	private static void String2Char(char[] buf, String str) {
		for (int k = 0; k < str.length(); ++k)
			buf[k] = str.charAt(k);
		buf[str.length()] = '\0';
	}
	private static boolean run(BufferedReader br) throws Exception {
		int q = Integer.parseInt(br.readLine());

		int id, grade, score;
		int cmd, ans, ret;
		boolean okay = false;

		for (int i = 0; i < q; ++i) {
			StringTokenizer st = new StringTokenizer(br.readLine(), " ");
			cmd = Integer.parseInt(st.nextToken());
			switch (cmd) {
				case CMD_INIT:
					usersolution.init();
					okay = true;
					break;
				case CMD_ADD:
					char[] gender = new char[7];
					id = Integer.parseInt(st.nextToken());
					grade = Integer.parseInt(st.nextToken());
					String2Char(gender, st.nextToken());
					score = Integer.parseInt(st.nextToken());
					ans = Integer.parseInt(st.nextToken());
					ret = usersolution.add(id, grade, gender, score);
					if (ret != ans)
						okay = false;
					break;
				case CMD_REMOVE:
					id = Integer.parseInt(st.nextToken());
					ans = Integer.parseInt(st.nextToken());
					ret = usersolution.remove(id);
					if (ret != ans)
						okay = false;
					break;
				case CMD_QUERY:
					int gradeCnt, genderCnt;
					int[] gradeArr = new int[3];
					char[][] genderArr = new char[2][7];
					gradeCnt = Integer.parseInt(st.nextToken());
					for (int j = 0; j < gradeCnt; ++j) {
						gradeArr[j] = Integer.parseInt(st.nextToken());
					}
					genderCnt = Integer.parseInt(st.nextToken());
					for (int j = 0; j < genderCnt; ++j) {
						String2Char(genderArr[j], st.nextToken());
					}
					score = Integer.parseInt(st.nextToken());
					ans = Integer.parseInt(st.nextToken());
					ret = usersolution.query(gradeCnt, gradeArr, genderCnt, genderArr, score);
					if (ret != ans)
						okay = false;
					break;
				default:
					okay = false;
					break;
			}
		}
		return okay;
	}

	public static void main(String[] args) throws Exception {
		int TC, MARK;

		//System.setIn(new java.io.FileInputStream("res/sample_input.txt"));

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine(), " ");

		TC = Integer.parseInt(st.nextToken());
		MARK = Integer.parseInt(st.nextToken());

		for (int testcase = 1; testcase <= TC; ++testcase) {
			int score = run(br) ? MARK : 0;
			System.out.println("#" + testcase + " " + score);
		}

		br.close();
	}
}

class student implements Comparable<student>{
	int id;
	int score;
	student(int id, int score) {
		this.id = id;
		this.score = score;
	}
	student() { }
	@Override
	public int compareTo(student arg0) {
		if (arg0.score != score) {
			return Integer.compare(score, arg0.score);
		} else {
			return Integer.compare(id, arg0.id);
		}
	}
}
class tuple {
	int grade, gender, score;
	tuple() { }
	tuple(int grade, int gender, int score) {
		this.grade = grade;
		this.gender = gender;
		this.score = score;
	}
}
class UserSolution {
	TreeSet<student>[][] st = new TreeSet[3][2];
	HashMap<Integer, tuple> tps;
	public void init() {
		for (int i = 0; i < 3; ++i) {
			for (int j = 0; j < 2; ++j) {
				st[i][j] = new TreeSet<>();
			}
		}
		tps = new HashMap<>();
	}

	public int add(int mId, int mGrade, char mGender[], int mScore) {
		--mGrade;
		st[mGrade][mGender[0] == 'f' ? 1 : 0].add(new student(mId, mScore));
		tps.put(mId, new tuple(mGrade, mGender[0] == 'f' ? 1 : 0, mScore));
		return st[mGrade][mGender[0] == 'f' ? 1 : 0].last().id;
	}

	public int remove(int mId) {
		if (tps.containsKey(mId) == false) return 0;
		tuple tp = tps.get(mId);
		st[tp.grade][tp.gender].remove(new student(mId, tp.score));
		tps.remove(mId);
		if (st[tp.grade][tp.gender].size() == 0) return 0;
		return st[tp.grade][tp.gender].first().id;
	}

	public int query(int mGradeCnt, int mGrade[], int mGenderCnt, char mGender[][], int mScore) {
		student ret = new student(0, (int)1e9);
		for (int i = 0; i < mGradeCnt; ++i) {
			--mGrade[i];
			for (int j = 0; j < mGenderCnt; ++j) {
				student stud = st[mGrade[i]][mGender[j][0] == 'f' ? 1 : 0].higher(new student(0, mScore));
				if (stud != null) {
					if (stud.compareTo(ret) < 0) {
						ret = stud;
					}
				}

			}
		}
		return ret.id;
	}
}